feat(ci): Support for install via CI

louisjimenez opened the pull request:
This adds support and provides documentation on how to install Spinnaker for GCP using Google Cloud Build or Jenkins. Using `setup.sh,` this allows a CI server to perform an equivalent installation or properties file update. Example configuration and documentation is added for getting started. This can eventually be migrated to Cloud docs.

louisjimenez commented on ci/README.md:
```diff
@@ -0,0 +1,141 @@
+
+These roles can be enabled through the IAM UI or with [gcloud](https://cloud.google.com/sdk/gcloud/reference/projects/add-iam-policy-binding).
+
+### Properties file
```
This can eventually be expanded to include additional detail on how to manage the properties file with Git and how to trigger builds

demobox commented on ci/README.md:
```diff
@@ -0,0 +1,141 @@
+- Service Usage Admin - roles/serviceusage.serviceUsageAdmin
+- Source Repository Administrator - roles/source.admin
+- Storage Admin - roles/storage.admin
+- Project IAM Admin - roles/resourcemanager.projectIamAdmin
```
A number of admin-level permissions here - do we know to what extent they're all needed?

louisjimenez replied:
Sure, I started with the least privilege and worked my way up as the setup ran into obstacles:

- Kubernetes Engine Admin: For creating a new cluster and for managing the API objects
- Cloud Memorystore Redis Admin: For creating a redis instance
- Service Usage Admin: For enabling required service apis
- Source Repository Administrator: For creating a new repo
- Storage Admin: For creating a new bucket
- Project IAM Admin: For getting the IAM roles on the service account executing the script and determining if the install can proceed

duftler commented on ci/CLOUD_BUILD.md:
```diff
@@ -0,0 +1,47 @@
+
+## Submitting a Build
+
+Cloud Builds can be trigged using [gcloud](https://cloud.google.com/cloud-build/docs/running-builds/start-build-manually), [build triggers](https://cloud.google.com/cloud-build/docs/running-builds/automate-builds), or [GitHub app triggers](https://cloud.google.com/cloud-build/docs/create-github-app-triggers). This repository provides the necessary files for installing Spinnaker for GCP with a gcl ...
```
s/trigged/triggered

duftler commented on ci/CLOUD_BUILD.md:
```diff
@@ -0,0 +1,47 @@
+
+## Service account
+
+Google Cloud Build has a default service account that can be found in the settings page of the Cloud Build UI. The Cloud Build service account must be configured with the following roles to install Spinnaker for GCP:
```
s/configured with/granted

duftler commented on scripts/install/setup.sh:
```diff
@@ -1,21 +1,75 @@
+
+source $REPO_PATH/spinnaker-for-gcp/scripts/manage/service_utils.sh
+
+REQUIRED_BINARIES=(git gcloud jq kubectl)
```
Could pull the required/missing binaries routine out into a function.

duftler commented on scripts/install/setup.sh:
```diff
@@ -1,21 +1,75 @@
 source "$PROPERTIES_FILE"
 
-~/spinnaker-for-gcp/scripts/manage/check_project_mismatch.sh
+if [ "$PROJECT_ID" != "$NETWORK_PROJECT" -a "$CI" = true ]; then
```
Same comment here about pulling this routine out into a function. Want to try to keep the size of the `setup.sh` file reasonable.

duftler commented on ci/install.bash:
```diff
@@ -0,0 +1,5 @@
+
+set -e
+
+REPO_PATH=/workspace PROPERTIES_FILE=/workspace/properties CI=true /workspace/spinnaker-for-gcp/scripts/install/setup.sh
```
Just a nit, but I expected `REPO_PATH` to be the full path to the repo, not the parent directory of the repo.

louisjimenez replied:
good point, going to change to `PARENT_DIR`

duftler commented on scripts/install/setup.sh:
```diff
@@ -1,21 +1,75 @@
 
-~/spinnaker-for-gcp/scripts/manage/check_duplicate_dirs.sh || exit 1
-~/spinnaker-for-gcp/scripts/manage/check_git_config.sh || exit 1
+[ -z "$REPO_PATH" ] && REPO_PATH="$HOME"
```
I find `[ -n "$REPO_PATH" ] || REPO_PATH="$HOME"` a bit more readable, but don't feel too strongly about it (if you think the style you've used is more idiomatic for this use case).

dorbin commented on ci/JENKINS.md:
```diff
@@ -0,0 +1,88 @@
+
+## Properties file
+
+In order to install Spinnaker for GCP, you must provide the installation script with a properties file. This properties file can be generated by executing [setup_properties.sh](../scripts/install/setup_properties.sh). Once you have a properties file, it must be provided to the installation script during the execution of the Jenkins job.
```
```suggestion
To get Jenkins to install Spinnaker for GCP, you need to generate a properties file and make it available to Jenkins:

1. Run [setup_properties.sh](../scripts/install/setup_properties.sh).
 
1. Make the resulting properties file available to the installation script as it executes the Jenkins job.
   You can use the [Credentials plug-in](https://wiki.jenkins.io/display/JENKINS/Credentials+Plugin), or use a secrets-management solution (like [Hashicorp Vault](https://www.vaultproject.io/) and configure Jenkins to read it from there.
```

dorbin commented on ci/JENKINS.md:
```diff
@@ -0,0 +1,88 @@
+
+In order to install Spinnaker for GCP, you must provide the installation script with a properties file. This properties file can be generated by executing [setup_properties.sh](../scripts/install/setup_properties.sh). Once you have a properties file, it must be provided to the installation script during the execution of the Jenkins job.
+
+### Jenkins Job Configuration
```
```suggestion
### Configure the Jenkins job
```

dorbin commented on ci/JENKINS.md:
```diff
@@ -0,0 +1,88 @@
+
+### Jenkins Job Configuration
+
+Once the aforementioned dependencies have been fulfilled, follow the below steps to configure a job for installing Spinnaker for GCP. 
```
```suggestion
Once the aforementioned dependencies have been fulfilled, follow the below steps to configure a job for installing Spinnaker for GCP. 
```
```suggestion
Once the dependencies are fulfilled, follow these steps to configure a job to install Spinnaker for GCP. 
```

dorbin commented on ci/JENKINS.md:
````diff
@@ -0,0 +1,88 @@
+REPO_PATH=$WORKSPACE PROPERTIES_FILE=$PROPERTIES CI=true $WORKSPACE/spinnaker-for-gcp/scripts/install/setup.sh
+```
+
+In the above example, the git `user.name` and `user.email` are configured prior to running `setup.sh`. This step is required for `setup.sh` to run successfully. Git operations can also be managed using the [Jenkins Git plugin](https://plugins.jenkins.io/git).
````
```suggestion
In the above example, the Git `user.name` and `user.email` must be configured before you run `setup.sh`. Git operations can also be managed using the [Jenkins Git plugin](https://plugins.jenkins.io/git).
```

dorbin commented on ci/JENKINS.md:
```diff
@@ -0,0 +1,88 @@
+- `PROPERTIES_FILE`: This is the absolute path to the your generated Spinnaker for GCP properties file.
+- `CI`: This must be set to `true` when running `setup.sh` outside of Cloud Shell.
+
+Once configured, execute the job to install Spinnaker for GCP. Changes to the properties file can be applied by re-running the job. Additional instructions for how to access or manage the deployed Spinnaker application are available [here](https://cloud.google.com/docs/ci-cd/spinnaker/spinnaker-for-gcp#access_spinnaker).
```
```suggestion
1. Execute the job to install Spinnaker for GCP. 

If you change the properties file, apply the change by re-running the job. 

Additional instructions for how to access or manage the deployed Spinnaker application are available [here](https://cloud.google.com/docs/ci-cd/spinnaker/spinnaker-for-gcp#access_spinnaker).
```

dorbin replied:
If you indent everything between step 4 and this step (Execute the job...), by three spaces, this becomes just another step in the procedure.

dorbin commented on ci/README.md:
```diff
@@ -0,0 +1,4 @@
+# Installing Spinnaker for GCP on a Continous Integration Server
+
+Spinnaker for GCP supports installation through a continous integration Server. A CI server can be used to conduct the initial installation and to apply updates when the Spinnaker for GCP properties file changes. Solutions for [Google Cloud Build](CLOUD_BUILD.md) and [Jenkins](JENKINS.md) are available.
```
```suggestion
You can install Spinnaker for GCP using a continous integration Server. A CI server can be used to conduct the initial installation and to apply updates when the Spinnaker for GCP properties file changes. Solutions for [Google Cloud Build](CLOUD_BUILD.md) and [Jenkins](JENKINS.md) are available.
```

louisjimenez commented on ci/JENKINS.md:
```diff
@@ -0,0 +1,88 @@
+
+These roles can be enabled through the IAM UI or with [gcloud](https://cloud.google.com/sdk/gcloud/reference/projects/add-iam-policy-binding).
+
+## Properties file
```
There are several different ways to do this. For the demo, I used the credentials plugin. What if I add some more detail on that:

*In this example, the properties file has been uploaded to the Jenkins server using the [Credentials plugin](https://support.cloudbees.com/hc/en-us/articles/203802500-Injecting-Secrets-into-Jenkins-Build-Jobs). It is made accessible to the job by binding the file to the PROPERTIES variable.*

dorbin replied:
That looks good, Louis. Is it worth mentioning what the other ways are?

"..has been uploaded to the Jenkins... You could also _____..."

louisjimenez replied:
I edited line 61 to call out the an alternative secrets management solution? Do you think that would be helpful?

dorbin replied:
:+1:

louisjimenez commented on ci/CLOUD_BUILD.md:
```diff
@@ -0,0 +1,47 @@
+
+Google Cloud Build has a default service account that can be found in the settings page of the Cloud Build UI. The Cloud Build service account must be configured with the following roles to install Spinnaker for GCP:
+
+- Cloud Functions Developer - roles/cloudfunctions.developer
```
It is a bit tricky because if the user is configuring through the UI they will be see the roles listed as the longform names, if they are adding them through the CLI they will need to add them using the role-paths

dorbin replied:
Makes sense. Let it stand.

louisjimenez commented on ci/CLOUD_BUILD.md:
```diff
@@ -0,0 +1,47 @@
+
+Cloud Builds can be trigged using [gcloud](https://cloud.google.com/cloud-build/docs/running-builds/start-build-manually), [build triggers](https://cloud.google.com/cloud-build/docs/running-builds/automate-builds), or [GitHub app triggers](https://cloud.google.com/cloud-build/docs/create-github-app-triggers). This repository provides the necessary files for installing Spinnaker for GCP with a gcl ...
+
+1. Create a new directory. The contents of this directory will be submitted to Cloud Build.
```
Anywhere in the shell is ok as long as the user runs the `gcloud builds submit` from the newly created directory. Should I add a note indicating the command should be run from there?

`Submit the build to Cloud Build from within the directory you previously created: gcloud builds submit --timeout "25m"  --config cloudbuild.yaml .`

dorbin replied:
Yes, that looks good

louisjimenez commented:
@duftler fyi for testing this on cloud build the git repo will need to point to my fork until it is merged. I modified the example cloudbuild.yaml below
```yaml
steps:
- name: 'gcr.io/cloud-builders/docker'
  args: ['build', '-f', 'Dockerfile', '-t', 'installer', '.']
- name: gcr.io/cloud-builders/git
  args: ['clone', 'https://github.com/louisjimenez/spinnaker-for-gcp.git']  
- name: gcr.io/cloud-builders/git
  args: ['config', '--global', 'user.name', '<example-user>']
- name: gcr.io/cloud-builders/git
  args: ['config', '--global', 'user.email', '<example-user@example.com>']
- name: gcr.io/cloud-builders/git
  args: ['checkout', 'louis/ci-install']
  dir: 'spinnaker-for-gcp'  
- name: 'installer'
  args: ['bash', './install.bash']
  env:
  - 'TERM=xterm'  
```
This is also the case for testing the Jenkins job, the cloned repo would need to point to the fork and checkout the branch:

```bash
#!/usr/bin/env bash

set -e

git clone https://github.com/louisjimenez/spinnaker-for-gcp.git
git config --global user.name "CI"
git config --global user.email "ci@example.com"

cd spinnaker-for-gcp
git checkout louis/ci-install

PARENT_DIR=$WORKSPACE PROPERTIES_FILE=$PROPERTIES CI=true $WORKSPACE/spinnaker-for-gcp/scripts/install/setup.sh

```

louisjimenez commented on ci/JENKINS.md:
```diff
@@ -0,0 +1,88 @@
+
+## Jenkins on GCP
+
+If your Jenkins server is running on GCP, follow [best practices](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#best_practices) for managing its service account. Your Jenkins server must have full access to all Google Cloud APIs to successfully install Spinnaker for GCP.
```
Yeah the instance is given full access to APIs but the actual IAM permissions are restricted by the IAM roles on its service account. That practice is lifted from the [best practices doc](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#best_practices). There is another doc that gives the steps for how to give an instance full Cloud API access, do you think it would be helpful to [link to it](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#changeserviceaccountandscopes) as well?

dorbin replied:
Yes, that sounds like a good idea.

louisjimenez commented:
@dorbin could you take one last look at this [commit](https://github.com/GoogleCloudPlatform/spinnaker-for-gcp/pull/135/commits/0393b1878dcdfd891c90ec8c405a0ede55d5c3f5)? I added a section on enabling the Cloud Resource API and added a link on how to modify instance access scopes. Thanks!

dorbin commented on ci/CLOUD_BUILD.md:
```diff
@@ -0,0 +1,57 @@
+
+## Enable Cloud Resource Manager API
+
+The Cloud Resource Manager API must be enabled for Cloud Build to successfully retrieve IAM policies. Enable it for your project by visiting the below URL and substituting your project number.
```
```suggestion
For Cloud Build to successfully retrieve IAM policies, you must enable the Cloud Resource Manager API. Visit this URL, substituting your project number.
```

dorbin commented on ci/JENKINS.md:
```diff
@@ -0,0 +1,102 @@
+
+## Jenkins on GCP
+
+If your Jenkins server is running on GCP, follow [best practices](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#best_practices) for managing its service account. Your Jenkins server must have full access to all Google Cloud APIs to successfully install Spinnaker for GCP. See the [Compute Engine documentation](https://cloud.google.com/compute/docs/access ...
```
:+1:

duftler commented on ci/JENKINS.md:
```diff
@@ -0,0 +1,102 @@
+`setup.sh` requires several variables to be passed in:
+
+- `PARENT_DIR`: The absolute path for the Jenkins workspace. Jenkins makes this available via `$WORKSPACE`.
+- `PROPERTIES_FILE`: This is the absolute path to the your generated Spinnaker for GCP properties file.
```
s/the your/your

duftler commented on scripts/install/setup.sh:
```diff
@@ -1,21 +1,55 @@
+
+PARENT_DIR=$PARENT_DIR PROPERTIES_FILE=$PROPERTIES_FILE $PARENT_DIR/spinnaker-for-gcp/scripts/manage/check_project_mismatch.sh
+
+if [ "$CI" = true ]; then
```
Is there a reason to only verify the operator service account has the required permissions when operating in "CI" mode?

louisjimenez replied:
I think you make a good point, the operator may be a cloud shell user who is not the project owner but would need these same permissions. I will update this.

duftler commented on scripts/install/setup.sh:
```diff
@@ -1,21 +1,55 @@
+PARENT_DIR=$PARENT_DIR PROPERTIES_FILE=$PROPERTIES_FILE $PARENT_DIR/spinnaker-for-gcp/scripts/manage/check_project_mismatch.sh
+
+if [ "$CI" = true ]; then
+  OPERATOR_SA_EMAIL=$(gcloud config list account --format "value(core.account)")
```
Need to explicitly specify `--project`.

duftler commented on ci/CLOUD_BUILD.md:
````diff
@@ -0,0 +1,57 @@
+```
+
+4. Copy the [Dockerfile](Dockerfile) and [install.bash](install.bash) file into the directory.
+5. Submit the build to Cloud Build: `gcloud builds submit --timeout "25m"  --config cloudbuild.yaml .`
````
Should probably explicitly specify `--project`.

duftler commented on scripts/install/setup.sh:
```diff
@@ -1,21 +1,55 @@
+if [ "$CI" = true ]; then
+  OPERATOR_SA_EMAIL=$(gcloud config list account --format "value(core.account)")
+
+  SETUP_REQUIRED_ROLES=(cloudfunctions.developer compute.networkViewer container.admin iam.serviceAccountCreator iam.serviceAccountUser pubsub.editor redis.admin serviceusage.serviceUsageAdmin source.admin storage.admin)
```
Having `roles/owner` would also be sufficient, yes?

duftler replied:
Not proposing we recommend this to the user, just pointing out that the script shouldn't fail if the service account has been granted just `roles/owner`.

louisjimenez replied:
Yeah `roles/owner` would work. And with the change to remove the `"$CI " = true` check from this  code block, it is more likely and operator will run into this issue. I'll update.

duftler commented on scripts/install/setup.sh:
```diff
@@ -1,21 +1,55 @@
+  MISSING_ROLES=""
+  for r in "${SETUP_REQUIRED_ROLES[@]}"; do
+    if [ -z "$(echo $SETUP_EXISTING_ROLES | grep $r)" ]; then
+      if [ -z $MISSING_ROLES ]; then
```
```suggestion
      if [ -z "$MISSING_ROLES" ]; then
```

duftler replied:
Otherwise, we get errors like these:

```
/home/duftler/spinnaker-for-gcp/scripts/install/setup.sh: line 40: [: cloudfunctions.developer,: binary operat
or expected
MISSING_ROLES=cloudfunctions.developer, compute.networkViewer, container.admin
/home/duftler/spinnaker-for-gcp/scripts/install/setup.sh: line 40: [: too many arguments
```

louisjimenez replied:
:+1:  Thank you

duftler commented on ci/CLOUD_BUILD.md:
```diff
@@ -0,0 +1,57 @@
+- Project IAM Admin - roles/resourcemanager.projectIamAdmin
+- Service Account User - roles/iam.serviceAccountUser
+
+You can enable these roles using the IAM UI or with [gcloud](https://cloud.google.com/sdk/gcloud/reference/projects/add-iam-policy-binding).
```
s/enable/grant

duftler commented on ci/CLOUD_BUILD.md:
```diff
@@ -0,0 +1,57 @@
+
+## Enable Cloud Resource Manager API
+
+For Cloud Build to successfully retrieve IAM policies, you must enable the Cloud Resource Manager API. Visit this URL, substituting your project number.
```
Could give guidance on where to locate project number.

duftler replied:
Actually, is this supposed to be "project id"?

louisjimenez replied:
Originally, I used the project number to enable/disable but both seem to work. I will update to use project id and make this consistent with the rest of the documentation.

duftler commented on ci/cloudbuild.yaml:
```diff
@@ -0,0 +1,13 @@
+- name: 'installer'
+  args: ['bash', './install.bash']
+  env:
+  - 'TERM=xterm'  
```
Extraneous trailing whitespace.

duftler commented on ci/JENKINS.md:
```diff
@@ -0,0 +1,102 @@
+
+1. Run [setup_properties.sh](../scripts/install/setup_properties.sh).
+ 
+1. Make the resulting properties file available to the installation script as it executes the Jenkins job. In this example, the properties file has been uploaded to the Jenkins server using the [Credentials plug-in](https://wiki.jenkins.io/display/JENKINS/Credentials+Plugin). It is made accessible to the job by binding the file to the PROPERTIES variable. Alternatively, you can use a secrets-mana ...
```
Missing closing parens after `Vault`.

duftler commented:
Did notice one thing in the jenkins testing that should be addressed (but could be done in a subsequent PR): When running the job for a second time, it fails with:

```
Started by user admin
Running as SYSTEM
Building in workspace /var/lib/jenkins/workspace/test
[test] $ /usr/bin/env bash /tmp/jenkins889045154286743382.sh
fatal: destination path 'spinnaker-for-gcp' already exists and is not an empty directory.
Build step 'Execute shell' marked build as failure
Finished: FAILURE
```
